src/snmp.c

```diff
@@ -1357,11 +1357,13 @@ static int csnmp_read_table(host_definition_t *host, data_definition_t *data) {
     if (oid_list_todo_num == 0) {
       /* The request is still empty - so we are finished */
       DEBUG("snmp plugin: all variables have left their subtree");
+      snmp_free_pdu(req);
       status = 0;
       break;
     }
 
     res = NULL;
+    /* snmp_sess_synch_response always frees our req PDU */
```

Comment:
>I think that req is (sometimes) freed when the return status is STAT_ERROR, otherwise we need to 
>free it ourselves.

Ok, we see some corner case when req is not freed when the return status from `snmp_sess_synch_response` is STAT_ERROR. That looks like error in net-snmp, and that practically should't happen, because session is opened.

Unpatched code of Collectd does not make a `snmp_free_pdu()` call when `status == STAT_ERROR`, so no regression added here by mine patch.

There are two calls of `snmp_sess_synch_response`

1) In `csnmp_read_table()`. 

Note what error handling code has comment "snmp_synch_response already freed our PDU".

```
  while (status == 0) {

    ...

    status = snmp_sess_synch_response(host->sess_handle, req, &res);
    if ((status != STAT_SUCCESS) || (res == NULL)) {
      ...
      if (res != NULL)
        snmp_free_pdu(res);
      res = NULL;

      /* snmp_synch_response already freed our PDU */
      req = NULL;        <<<---------------------
      ...

      status = -1;
      break;
    }
    ...
  }
....
  if (req != NULL)
    snmp_free_pdu(req);
```
No `snmp_free_pdu(req)` will be called in case of `snmp_sess_synch_response` errors.

2) In `csnmp_read_value()`. 

That implementation uses the same `snmp_pdu_create() + snmp_sess_synch_response()` chain of calls, and there is no `snmp_free_pdu(req)` at all.